src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs

```diff
@@ -1326,6 +1345,237 @@ static void Main()
+{
+    static void Main()
+    {
+        int VARIABLES; // int x1, x2, ...
+        (VARIABLES) = (TUPLE).ToTuple(); // (x1, x2, ...) = (1, 2, ...).ToTuple();
+
+        System.Console.WriteLine(OUTPUT);
+    }
+}
+";
+            for (int i = 2; i <= 21; i++)
+            {
+                var tuple = String.Join(", ", Enumerable.Range(1, i).Select(n => n.ToString()));
+                var variables = String.Join(", ", Enumerable.Range(1, i).Select(n => $"x{n}"));
+                var output = String.Join(@" + "" "" + ", Enumerable.Range(1, i).Select(n => $"x{n}"));
+                var expected = String.Join(" ", Enumerable.Range(1, i).Select(n => n));
+
+                var source = template.Replace("VARIABLES", variables).Replace("TUPLE", tuple).Replace("OUTPUT", output);
+                var comp = CompileAndVerify(source, expectedOutput: expected, additionalRefs: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature().WithRefsFeature());
+                comp.VerifyDiagnostics();
+            }
+        }
+
+        [Fact]
+        public void CannotDeconstructRefTuple22()
+        {
+            string template = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        int VARIABLES; // int x1, x2, ...
+        (VARIABLES) = CreateLongRef(1, 2, 3, 4, 5, 6, 7, CreateLongRef(8, 9, 10, 11, 12, 13, 14, Tuple.Create(15, 16, 17, 18, 19, 20, 21, 22)));
+    }
+
+    public static Tuple<T1, T2, T3, T4, T5, T6, T7, TRest> CreateLongRef<T1, T2, T3, T4, T5, T6, T7, TRest>(T1 item1, T2 item2, T3 item3, T4 item4, T5 item5, T6 item6, T7 item7, TRest rest) =>
+        new Tuple<T1, T2, T3, T4, T5, T6, T7, TRest>(item1, item2, item3, item4, item5, item6, item7, rest);
+}
+";
+            var tuple = String.Join(", ", Enumerable.Range(1, 22).Select(n => n.ToString()));
+            var variables = String.Join(", ", Enumerable.Range(1, 22).Select(n => $"x{n}"));
+
+            var source = template.Replace("VARIABLES", variables).Replace("TUPLE", tuple);
+
+            var comp = CreateCompilationWithMscorlib(source, references: new[] { ValueTupleRef, SystemRuntimeFacadeRef }, parseOptions: TestOptions.Regular.WithTuplesFeature());
+            comp.VerifyDiagnostics(
+                // (8,113): error CS1501: No overload for method 'Deconstruct' takes 22 arguments
+                //         (x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12, x13, x14, x15, x16, x17, x18, x19, x20, x21, x22) = CreateLongRef(1, 2, 3, 4, 5, 6, 7, CreateLongRef(8, 9, 10, 11, 12, 13, 14, Tuple.Create(15, 16, 17, 18, 19, 20, 21, 22)));
+                Diagnostic(ErrorCode.ERR_BadArgCount, "CreateLongRef(1, 2, 3, 4, 5, 6, 7, CreateLongRef(8, 9, 10, 11, 12, 13, 14, Tuple.Create(15, 16, 17, 18, 19, 20, 21, 22)))").WithArguments("Deconstruct", "22").WithLocation(8, 113)
+                );
+        }
+
+        [Fact]
+        public void DeconstructUsingDynamicMethod()
+        {
+            string source = @"
```

Comment:
Yes. The "Dynamic" test covers the scenario (dynamic types on the left and as output parameters):

```C#
class C
{
    dynamic Dynamic1;
    dynamic Dynamic2;

    static void Main()
    {
        C c = new C();
        (c.Dynamic1, c.Dynamic2) = c;
        System.Console.WriteLine(c.Dynamic1 + "" "" + c.Dynamic2);
    }

    public void Deconstruct(out int a, out dynamic b)
    {
        a = 1;
        b = ""hello"";
    }
}
```